File: Data Structures & Algorithms/maximum-product-subarray/submission-1.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int prod = 1;
        int prod2 = 1;
        int maxProd = nums[0];

        int negCount = 0;

        for (int num:nums) {
            prod *= num;
            maxProd = max(prod, maxProd);

            if (negCount > 0) {
                prod2 *= num;
                maxProd = max(maxProd, prod2);
            }
            if (num < 0) negCount++;

            if (num == 0) {
                negCount = 0;
                prod = 1;
                prod2 = 1;
            }

        }

        return maxProd;
    }
};

File: Data Structures & Algorithms/perfect-squares/submission-1.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [n]*(n+1)
        dp[0] = 0
        for num in range(1, n+1):
            i = 1
            while i ** 2 <= num:
                dp[num] = min(dp[num], 1 + dp[num-i**2])
                i += 1

        return dp[n]